ast/compile_test.go

```diff
@@ -1951,6 +1951,42 @@ func TestCompilerCheckRuleConflictsDotsInRuleHeads(t *testing.T) {
 				`),
 			err: "rego_type_error: single-value rule data.pkg.p.q[r] conflicts with [data.pkg.p.q.r.s]",
 		},
+		{
+			note: "single-value partial object with other partial object rule overlap, unknown keys (regression test for #5855)",
+			modules: modules(
+				`package pkg
+				p[r] := x { r = input.key; x = input.bar }
+				p.q[r] := x { r = input.key; x = input.bar }
+				`),
+			err: "rego_type_error: single-value rule data.pkg.p[r] conflicts with [data.pkg.p.q[r]]",
+		},
+		{
+			note: "single-value partial object with other partial object (implicit 'true' value) rule overlap, unknown keys",
+			modules: modules(
+				`package pkg
+				p[r] := x { r = input.key; x = input.bar }
+				p.q[r] { r = input.key }
+				`),
+			err: "rego_type_error: single-value rule data.pkg.p[r] conflicts with [data.pkg.p.q[r]]",
+		},
+		{
+			note: "single-value partial object with multi-value rule (ref head) overlap, unknown key",
+			modules: modules(
+				`package pkg
+				import future.keywords
+				p[r] := x { r = input.key; x = input.bar }
+				p.q contains r { r = input.key }
+				`),
```

Comment:
this intentionally is not an error?

Reply:
Yes. This should be valid, right?
The first rule is building an object, and the second rule is assigning a set to the key `q` in that object.

Of course, I could also imagine that we want to reject this at compile-time, with the argument that you're only allowed to make single-value rules that overlap a partial object rule. E.g. the above could be achieved with a comprehension: `p.q := [r | r := input.key]`.

Reply:
For what it's worth, this will trigger [this issue](https://github.com/open-policy-agent/opa/issues/5859) when evaluated. The comprehension alternative will not.

Reply:
Yes, that'll trigger a `eval_conflict_error: object keys must be unique`